### DISCUSSION TITLE
Activities not seen in Grafana

### DISCUSSION THREAD

**Discussion by IrfanNian** [cmu-sei/GHOSTS#207] [Q&A] [answered]
Hi, first of all, such a great project! I am an InfoSec undergraduate and am planning to make use of GHOSTS on one of my side projects. Can say that I am pretty unfamiliar with some of the technologies used here. So to start off, as do some people here, I followed the guide from this website - https://cylab.be/blog/71/simulate-user-activity-with-the-ghosts-framework-introduction 

My setup is - Ubuntu VM as Server, Win11 VM as Client. Managed to get the basics up and running -
`docker ps` on Ubuntu
<img width="719" alt="image" src="https://user-images.githubusercontent.com/63865160/231714081-6f7547cb-f55c-4bbd-b08f-4c5bab0dbb0c.png">

Grafana managed to detect the Win11 Client successfully
<img width="641" alt="image" src="https://user-images.githubusercontent.com/63865160/231714447-d0da751c-a532-4a2a-a00f-f2e742541f4f.png">

**However**, the issue I faced is that when using the default timeline.json and modify application.json accordingly (changed the IPs accordingly, using the Ubuntu IP)
<img width="331" alt="image" src="https://user-images.githubusercontent.com/63865160/231715313-59594680-eab5-417e-8caf-e2d1b54a5bdc.png">

And upon running `ghost.exe` , in addition to downloading the latest Chrome & Firefox and also Office 2019 (Unlicensed)
<img width="594" alt="image" src="https://user-images.githubusercontent.com/63865160/231715631-500a6368-d577-4e74-8a4c-616804ea4108.png">

I am not able to see any Activities in Grafana. On my Win11, I can see it created a Word document and writes something, Excel just opens and close no xlsx files created, blank Powerpoint is created. Firefox and Chrome didn't even launch to browse any websites.
<img width="642" alt="image" src="https://user-images.githubusercontent.com/63865160/231718406-4e76c98a-43db-4984-b668-7ce3d194b806.png">

Any thoughts on why this happens? 

Tried to also access the Swagger page on my Win11, this is what I get when trying the `POST /api/clientresults`
<img width="764" alt="image" src="https://user-images.githubusercontent.com/63865160/231727980-1d301379-04b3-4f5a-afc7-345027770fa8.png"> Could this be the issue?

Also, for Office, does it being Unlicensed cause any issues? Because when it launches, the product key panel will appear automatically.

**Comment by sei-dupdyke** [comment]
Hi, thanks for checking out the project.

Sometimes the quick fix is to go into the gear menu at left, and go to datasources. Click through and see if the connection is successful.

**Reply by IrfanNian** [reply]
Hi,

Upon checking, it seems to be fine.
<img width="566" alt="image" src="https://user-images.githubusercontent.com/63865160/231927759-f9c3a661-0562-4d5b-a987-8e41827535b0.png">

Also, any reasons why Chrome and Firefox not browsing anything despite using default timeline.json? Cos it be due to the browsers being the latest version?

**Reply by sei-dupdyke** [reply]
Do you have the time shown set to all? It's in the upper right-hand corner:
![image](https://user-images.githubusercontent.com/43444464/232033173-0fb34d27-3442-4479-a1f6-420296531bc3.png)

For the browsers, do you have the appropriate versioned chrome and gecko drivers installed in the same directory as ghosts? Check the logs/application.log for errors on this.

**Reply by IrfanNian** [reply]
Yes, I have changed it to even 'This Fiscal Year' and still Activities are showing zero. Just the Polling number keeps increasing. 

For the browser, seems like I found the culprit for Chrome as shown below -
<img width="624" alt="image" src="https://user-images.githubusercontent.com/63865160/232371782-7b8998a7-7203-4279-96e4-9f91e7beb3cf.png">

Also one quick question, does an unlicensed MS Office prevents GHOSTS from working as intended, ie. activities listed in the default timeline.json? Using unlicensed/unactivated due to me still being a student. If it does affects, are there other alternative you would suggest that works for GHOSTS ie. WPS / Open Office?

**Reply by sei-dupdyke** [reply]
Hmm, so let's see if history is in your database.

docker exec -it ghosts-postgres /bin/bash

psql -U ghosts

select count(*) from history_timelines;

For the chrome error, you have to download the appropriate version of chromedriver as outlined here: https://cmu-sei.github.io/GHOSTS/core/client/

Unlicensed office throws lots of errors, I don't recommend it. I haven't had time to work on open office integration, sorry.

**Reply by IrfanNian** [reply]
This was what I got - 
<img width="364" alt="image" src="https://user-images.githubusercontent.com/63865160/233259259-bac81d85-a480-4f30-834d-3aea0f2f6d4b.png">

Oh, i see, no worries, I've found a workaround for Office licensing. In addition, when I launched ghost.exe, in the app logs, this error was found, may I know if it affects anything? -
`2023/04/19 06:11:38.659|Ghosts.Client.Handlers.BaseBrowserHandler.ExecuteEvents|OpenQA.Selenium.WebDriverException: Could not connect to the remote WebDriver server for URL http://localhost:64528/session/497fc8a5e5cd379dc28f69fc9f3243d1/window.`

Also, I have additional questions regarding GHOSTS capability, would appreciate it if you could answer them - _Is GHOSTS capable of doing the stuff below?_

**1. Microsoft Outlook**
- Login to Email
- Download Attachments

**2. Chrome/Firefox**
- Login to Websites (ie. Facebook, Twitter)

**3. Windows File Explorer**
- Create New Folder/File
- Rename Folder/File
- Delete Folder/File
- Copy & Paste Folder/File from one location to another
Save Folder/File

**4. Software Installation/Clickable** 
- Installing software by clicking the downloaded .exe file (through drive-by-download) in Downloads folder and clicking through 'Next/Finish' in dialog prompt when trying to install

**Reply by sei-dupdyke** [reply, marked as answer]
Sorry it is `select * from history_timeline;` - drop the "s"

**Reply by IrfanNian** [reply]
Hi, sorry for the late reply. Its finally showing the activities, didn't change any configuration. Probably took awhile to display it.
<img width="607" alt="image" src="https://github.com/cmu-sei/GHOSTS/assets/63865160/44fea57c-3efe-4031-a421-f52e7edbe7f7">

However, seems like Health is still showing '0'. Am I supposed to expect something to be shown in Health?

In addition, are you able to answer the additional questions asked above, especially _1. Microsoft Outlook_ and _4. Software Installation/Clickable_ as these tasks (ie. User logging in to Outlook at a certain time, User 'clicking' through installation setup window) are crucial in the project I'm doing now in regards to user emulation

Thank you.

**Reply by sei-dupdyke** [reply]
Good to hear!

Health might be turned off on the agents? We changed this to be off by default beginning around v6 or so?

On the client, check ./config/application.json -> "HealthIsEnabled"

https://github.com/cmu-sei/GHOSTS/blob/master/src/Ghosts.Client/config/application.json#LL43C23-L43C23

**Reply by IrfanNian** [reply]
Hi, yes! That worked! However after generating over 100k+ activities, suddenly one fine day, my entire Dashboard went empty. Everything was wiped out. Changed to 'Last 1 year' also didn't change anything. Could this be due me adding a Firewall to my Simulated Environment? But shouldn't the past activities still be displayed regardless of incoming traffic being block? That being said, I allowed inbound of Port 3000,5000, 5432, is there anymore ports I should take note of? And is there anyway for me to retrieve back my past generated data?

**Reply by sei-dupdyke** [reply]
Is the data still there?

```
docker exec -it ghosts-postgres /bin/bash

psql -U ghosts

select count(*) from history_timelines;
```

I can't imagine how a firewall rule affects this.

**Reply by IrfanNian** [reply]
Sflr. Yes its apparently still there. Unsure why it suddenly was removed from the Dashboard, but no worries, not a big issue.

Additionally I want to ask - I know that GHOSTS doesn't support any open-source Office (ie. Open Office) or email clients (ie. EM Client), but is GHOSTS able to do basic stuffs like creating new OO files, closing, deleting OO files and also do simple email sending using EM Client? I saw you mentioning about LightHandlers in the another Discussion post, however I can't seem to find a template/guide on how to use it. 

One more thing - does 'Click' able to do clicking for eg. clicking 'Next' button on an program installation Box Dialog? As I want to mimic a user installing programme.

Thanks.

**Reply by sei-dupdyke** [reply]
Correct, there is currently no support for OpenOffice - I just don't have time to develop this, and I haven't had a lot of people asking for that feature. LightHandlers just create files without using Office, etc. and that is probably of limited use for most people.

Click is not for clicking on specific things, but rather for randomly clicking areas of the screen so as to get around apps that go to sleep, screen locking, etc.

**Reply by IrfanNian** [reply]
I see. So, if I have OpenOffice and EM Client in my environment right now, there's **no way** for GHOSTS to do basic launching of OO Writers (Word) / Calc (Excel) .exe and to launch, save, delete (don't need modify) the files and do basic email sending function on EM Client?

Ahh, noted on 'Click'. So as of now no way to mimic user installing programme as typical installation of programme requires interaction with the Dialog Box for installation location, etc?

Thanks.

**Reply by sei-dupdyke** [reply]
For Open Office, you could possibly use [the command line](https://wiki.openoffice.org/wiki/Documentation/OOoAuthors_User_Manual/Getting_Started/Starting_from_the_command_line) to open it that way and try to generate content, but I have not tried this and can't guarantee it will work (nor be a pain to use that way). Sorry.

**Reply by IrfanNian** [reply]
Ahh thanks for the pointer!

Also, in timeline.json, how do I make the activities (ie. Send emails, visit websites, etc. ) execute every several hours or even better - randomly throughout between the set UtcTimeOn and UtcTimeOff timing? Do I just have to set "Loop" to true?
